### PR TITLE
Issue 240 base detector

### LINKED ISSUE #240: Implement BaseDetector Pipeline and EPR/WEPR aliases
Implement the `BaseDetector` pipeline class to replace the standalone legacy detectors.

## Implementation decision
Instead of writing bespoke logic for each reduction (EPR / WEPR), `BaseDetector` inherits directly from `sklearn.pipeline.Pipeline`. This standardizes the entire feature extraction and classification process. 
- Sequence-level classification is handled automatically by the native `.predict_proba()` method inherited from `Pipeline` (we do **not** override it — design decision #2).
- Token-level classification is handled by a custom `.predict_token_proba()` method.

## Construction (two paths)
- `EPR(**kwargs)` / `WEPR(**kwargs)` build an **unfitted** detector (a fresh `PretrainedLogisticRegression`) — for calibration training via `.fit()` (#282). Calling `predict*` before fitting raises `NotFitted`.
- `BaseDetector.from_pretrained(model, reduction)` builds a **ready-to-predict** detector with weights loaded (#279).

Both share the same step structure; the transformers are stateless (`requires_fit = False`, #277/#278), so only the classifier carries fitted state.

## Two prediction paths
Both methods share the parser step, then diverge at `EntropyTransformer`: `predict_proba` reduces over tokens (one score per sequence), `predict_token_proba` keeps one score per token.

```mermaid
flowchart TD
    X["X — raw outputs"] --> P["LogProbParser.transform<br/>(n, max_tokens, k)"]
    subgraph seq["predict_proba — sequence level"]
        direction TB
        SA["EntropyTransformer.transform<br/>reduce over tokens<br/>(n, n_features)"] --> SC["classifier.predict_proba<br/>→ (n, 2); [:, 1] = P(halluc.)"]
    end
    subgraph tok["predict_token_proba — token level"]
        direction TB
        TA["EntropyTransformer.transform_tokens<br/>per token, no reduction<br/>(n, max_tokens, n_features)"] --> TC["flatten → classifier.predict_proba → reshape<br/>→ (n, max_tokens, 1) one score per token"]
    end
    P --> SA
    P --> TA
```

## Algorithm for predict_token_proba
`predict_token_proba(X)` manually routes data through the pipeline in token mode, then **masks padded tokens before the classifier** (which rejects NaN):
1. **Apply transformations (token mode):** Iterate `self._iter(with_final=False)`. For each transformer, call `transform_tokens(Xt)` if present, else fall back to `transform(Xt)`. The output is a padded 3D array `(n_samples, max_tokens, n_features)` with **`NaN` at padded token positions**.
2. **Flatten:** reshape to `(n_samples * max_tokens, n_features)`.
3. **Mask padded rows:** `valid = ~np.isnan(rows).any(axis=1)`. This is required — `LogisticRegression.predict_proba` runs `check_array(ensure_all_finite=True)` and raises `ValueError` on any `NaN` row, which would otherwise hit on every variable-length batch.
4. **Predict on valid rows only:** `classifier = self.steps[-1][1]`; `p = classifier.predict_proba(rows[valid])[:, 1]` (positive class = hallucination).
5. **Scatter + reshape:** prefill a `(n_samples * max_tokens,)` output with `NaN`, set `out[valid] = p`, reshape to `(n_samples, max_tokens, 1)`. Padded positions stay `NaN` (there is no token there).

## Empty sequences
A completion with zero tokens produces an all-`NaN` feature row. Since `predict_proba` is the inherited (non-overridden) `Pipeline.predict_proba`, the NaN must be resolved **upstream**: `EntropyTransformer.transform` (#277) maps an all-`NaN` (token-less) sequence to a **zero feature vector**, so the classifier returns the calibrated baseline `sigmoid(intercept_)` (matches legacy). For `predict_token_proba`, an empty sequence's token rows are all masked out → that sequence's tokens are all `NaN` in the output.

## Skeleton (signatures & contracts — bodies to be implemented)
```python
from sklearn.pipeline import Pipeline
import numpy as np

class BaseDetector(Pipeline):
    def predict_token_proba(self, X) -> np.ndarray:
        """Token-level hallucination probabilities, shape (n_samples, max_tokens, 1).

        Route the pipeline in token mode (transform_tokens, with transform fallback),
        mask padded (NaN) token rows before the classifier, scatter results back so
        padded positions stay NaN. See "Algorithm for predict_token_proba".
        """
        raise NotImplementedError

    @classmethod
    def from_pretrained(cls, pretrained_model_name_or_path: str, reduction: str) -> "BaseDetector":
        """Build a ready-to-predict detector with weights loaded (#279). See acceptance criteria."""
        raise NotImplementedError


# Factory aliases (module level, exposed via #237) — each returns a BaseDetector wired as
#   LogProbParser -> EntropyTransformer(reduction="epr"|"wepr") -> PretrainedLogisticRegression
def EPR(**kwargs) -> "BaseDetector":
    raise NotImplementedError

def WEPR(**kwargs) -> "BaseDetector":
    raise NotImplementedError
```

## Acceptance criteria
- [x] Inherits directly from `sklearn.pipeline.Pipeline`.
- [x] Factory functions `EPR(**kwargs)` and `WEPR(**kwargs)` return a `BaseDetector` with steps `LogProbParser` → `EntropyTransformer(reduction="epr")` / `EntropyTransformer(reduction="wepr")` → `PretrainedLogisticRegression` (the `reduction` keyword matches #277; classifier per #279).
- [x] `predict_token_proba(X)` routes in token mode and **masks padded (`NaN`) token rows** before `classifier.predict_proba`, scattering results back so padded positions stay `NaN`; output `(n_samples, max_tokens, 1)`. Verified on a variable-length batch (different sequence lengths).
- [x] Empty (token-less) sequences are handled: `predict_proba` returns the baseline (via the zero-feature mapping in #277), `predict_token_proba` returns all-`NaN` tokens for that sequence.
- [x] `from_pretrained(cls, pretrained_model_name_or_path, reduction)` builds and returns a ready-to-predict `BaseDetector` — steps `LogProbParser`, `EntropyTransformer(reduction=...)`, `PretrainedLogisticRegression.from_pretrained(...)` (#279) — propagating loaded hyperparameters (e.g. `k`). No `fit` call is needed: the transformers are stateless (`requires_fit = False`, #277/#278) and the classifier is pre-fitted.

### PR DESCRIPTION
**Pull request by CharlotteLeBihan18**
closes #240

### CODE REVIEW

File: src/artefactual/scoring/base_detector.py [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+        Routes the pipeline in token mode (transform_tokens, with transform fallback),
+        mask padded (NaN) token rows before the classifier, scatter results back so
+        padded positions stay NaN.
+        """
+        raw_output = x
+        for _, transformer in self.steps[:-1]:
+            try:
+                raw_output = transformer.transform_tokens(raw_output)
+            except AttributeError:
+                raw_output = transformer.transform(raw_output)
```
Comment (chicham) [inline src/artefactual/scoring/base_detector.py]: The `try` here spans the *execution* of `transform_tokens`, so an `AttributeError` raised inside a step's `transform_tokens` implementation is silently swallowed and the step is rerouted to `transform` — the token path then continues with sequence-reduced, wrong-shaped data. Narrow the `try` to the attribute lookup so only "method doesn't exist" triggers the fallback:

```suggestion
            try:
                step_transform = transformer.transform_tokens
            except AttributeError:
                step_transform = transformer.transform
            raw_output = step_transform(raw_output)
```

Longer term, a structural `TokenTransformer` Protocol checked with beartype would remove this dispatch entirely.

File: src/artefactual/scoring/entropy_methods/entropy_contributions.py [hunk 2]
```diff
@@ -15,10 +15,11 @@ class EntropyContributionsMixin:
     def entropy_contributions(logprobs: NDArray[np.floating]) -> NDArray[np.floating]:
         """Compute entropic contributions s_kj = -p_k log(p_k) for top-K logprobs using vectorized operations.
         Args:
-            logprobs: An array of log probabilities, the last axis being the per-token rank axis.
+            logprobs: A 2D array of shape (num_tokens, num_logprobs) containing log probabilities.
+            k: Number of top log probabilities to consider per token.
```
Comment (chicham) [inline src/artefactual/scoring/entropy_methods/entropy_contributions.py]: This docstring edit reintroduces a `k:` parameter description, but the function signature has no `k` — main's current docstring was just fixed to remove exactly this (60242eb). Also "A 2D array of shape (num_tokens, num_logprobs)" narrows the contract: the function is dimension-agnostic (it is called with 3D batches from `EntropyTransformer`); main's wording ("last axis being the per-token rank axis") is the accurate one. Suggest dropping this docstring change from the PR.

**Review by chicham — COMMENTED** [review]
@chicham — merge-process note:

Do NOT squash — squashing the previous PR is what created the duplicate-commit conflicts this branch just had to merge its way out of. Use a rebase merge, after first rewriting the branch so it carries only its own work on top of current main (right now it still holds duplicates of the previously squash-merged PR's commits plus a merge commit, which rebase-merge refuses to replay):

```bash
git fetch origin
git checkout -B issue-240-Base-Detector origin/issue-240-Base-Detector
git reset --soft origin/main   # keep the final tree, drop the tangled history
git commit -m "feat: add BaseDetector pipeline with epr/wepr factories and token-level prediction (#240)"
git push --force-with-lease origin issue-240-Base-Detector
gh pr merge 291 --rebase
```

(The current head's tree is already correct — it merged main and resolved conflicts the right way — so resetting softly onto main and recommitting loses nothing. Split into several commits before pushing if you want finer history.)

Rebase-merging replays the branch's commits onto main with identical patches, so the still-open stacked branches (issue-282 next) can rebase onto the new main and their duplicated commits drop out automatically — no add/add conflicts like this PR hit.

**Review by chicham — DISMISSED** [review]
[NOTE] @chicham — merge-process note:

The latest head merges main into the branch, resolving the earlier conflict correctly: pretrained_regression.py and utils/io.py now match main exactly (#290's load_weights registry + ValueError guard survive), and tests/scoring/test_base_detector.py was added (properly collected, real data, no mocks). True diff is 4 files: base_detector.py, test_base_detector.py, and the two entropy files.

Verified earlier on an equivalent rebased state: pre-commit 16/16, pytest green, plus manual end-to-end checks of predict_proba / predict_token_proba (padding-NaN masking, empty-sequence baseline sigmoid(intercept_)) matching the #240 acceptance criteria. Re-run pre-commit + pytest on this exact head before merging.

Ready to merge when: (1) the inline comments are addressed or explicitly deferred — the EntropyTransformer k/_align_to_k removal is tracked in #296; (2) checks are green on this head. Use SQUASH — rebase-merge will refuse: the branch still carries #290-era duplicate commits plus a merge commit. After landing, rebase issue-282 (#292) onto the new main before its review.

Known gaps, non-blocking (tracked): WEPR end-to-end is untested (WEPR_WEIGHTS defined but unused in test_base_detector.py — likely blocked by the #296 width question); overall coverage tracked in #295/#245.
